numpy/lib/function_base.py

```diff
@@ -2669,6 +2689,29 @@ class vectorize(object):
            <http://docs.scipy.org/doc/numpy/reference/c-api.generalized-ufuncs.html>`_.
     """
 
+    # __new__ is overriden as a clean way to provide decorator-with-keywords
+    # syntax while allowing 'vectorize' to remain as the type of vectorized
+    # functions.
+    def __new__(cls, pyfunc=None, *args, **kwargs):
+
+        # For backwards compatibility, subclasses must not use the overloaded
+        # __new__, as vectorize.__new__ and subclass.__init__ will receive the
+        # same arguments.
+        if cls is not vectorize:
```

Comment:
> (I think doing introspection on __init__ for the name of the first argument probably goes a bit too far...)

It would also be inadequate, since a subclass could easily put the argument corresponding to `pyfunc` as e.g. the second positional arg in `__init__`. (and in fact, my test does this)